### PR TITLE
IOS-705: Blocking users mechanism.

### PR DESCRIPTION
**Pull request by pbruz**
### Ticket
[IOS-705](https://netguru.atlassian.net/browse/IOS-705)

### Task Description
Blocking users mechanism.

### CODE REVIEW

File: Inbbbox/Source Files/ViewControllers/SettingsViewController.swift [hunk 1]
```diff
@@ -172,8 +172,15 @@ extension SettingsViewController {
             labelItem.onSelect?()
         }
 
-        if item is DetailsItem {
-            navigationController?.pushViewController(LanguageViewController(), animated: true)
+        if let detailsItem = item as? DetailsItem {
+            guard let destination = detailsItem.destination else { return }
```
Comment (bwdowikowski) [inline Inbbbox/Source Files/ViewControllers/SettingsViewController.swift]: Hmm.. my proposition about this `if-let`

```swift
if let destination = (item as? DetailsItem).destination {
    let destinationViewController: UIViewController = {
        switch destination {
        case .language: return LanguageViewController()
        case .blockedUsers: return BlockedUsersViewController()
        }
    }()
    navigationController?.pushViewController(destinationViewController, animated: true)
}
```

File: Inbbbox/Source Files/ViewModels/BucketsViewModel.swift [hunk 2]
```diff
@@ -94,20 +94,25 @@ class BucketsViewModel: BaseCollectionViewViewModel {
...
-                }
-
-                if bucketShotsShouldBeReloaded {
-                    let indexPath = IndexPath(row: index, section: 0)
-                    self.delegate?.viewModel(self, didLoadShotsForItemAtIndexPath: indexPath)
+                if let unwrappedShots = shots {
+                    firstly {
+                        self.removeShotsFromBlockedUsers(unwrappedShots)
+                    }.then { filteredShots -> Void in
+                        if let oldShots = self.bucketsIndexedShots[index], let newShots = filteredShots {
```
Comment (bwdowikowski) [inline Inbbbox/Source Files/ViewModels/BucketsViewModel.swift]: There should be new line after a comma.

File: Inbbbox/Source Files/ViewModels/BucketsViewModel.swift [hunk 3]
```diff
@@ -151,9 +173,8 @@ extension BucketsViewModel {
 
         init(bucket: BucketType, shots: [ShotType]?) {
             self.name = bucket.name
-            self.numberOfShots = String.localizedStringWithFormat(Localized("%d shots",
-                    comment: "How many shots in collection?"), bucket.shotsCount)
             if let shots = shots, shots.count > 0 {
+                self.numberOfShots = String.localizedStringWithFormat(Localized("%d shots", comment: "How many shots in collection?"), shots.count)
```
Comment (bwdowikowski) [inline Inbbbox/Source Files/ViewModels/BucketsViewModel.swift]: Isn't `self` is redundant here?

File: Inbbbox/Source Files/ViewModels/BucketsViewModel.swift [hunk 4]
```diff
@@ -166,6 +187,7 @@ extension BucketsViewModel {
                 }
             } else {
                 self.shotsImagesURLs = nil
+                self.numberOfShots = String.localizedStringWithFormat(Localized("%d shots", comment: "How many shots in collection?"), 0)
```
Comment (bwdowikowski) [inline Inbbbox/Source Files/ViewModels/BucketsViewModel.swift]: Also here?

File: Inbbbox/Source Files/ViewModels/FolloweesViewModel.swift [hunk 5]
```diff
@@ -37,9 +37,17 @@ class FolloweesViewModel: BaseCollectionViewViewModel {
...
         }.then { followees -> Void in
-            if let followees = followees, followees != self.followees || followees.count == 0 {
-                self.followees = followees
-                self.delegate?.viewModelDidLoadInitialItems()
+            if let followees = followees {
+
+                firstly {
+                    self.removeBlockedUsers(followees)
+                }.then { filteredFollowees -> Void in
+                    if let filteredFollowees = filteredFollowees, filteredFollowees != self.followees || filteredFollowees.count == 0 {
```
Comment (bwdowikowski) [inline Inbbbox/Source Files/ViewModels/FolloweesViewModel.swift]: Should be new line after a comma.

File: Inbbbox/Source Files/ViewModels/LikesViewModel.swift [hunk 6]
```diff
@@ -30,9 +30,17 @@ class LikesViewModel: SimpleShotsViewModel {
...
         }.then { shots -> Void in
-            if let shots = shots?.map({ $0.shot }), shots != self.shots || shots.count == 0 {
-                self.shots = shots
-                self.delegate?.viewModelDidLoadInitialItems()
+            if let shots = shots?.map({ $0.shot }) {
+
+                firstly {
+                    self.removeShotsFromBlockedUsers(shots)
+                }.then { filteredShots -> Void in
+                    if let filteredShots = filteredShots, filteredShots != self.shots || filteredShots.count == 0 {
```
Comment (bwdowikowski) [inline Inbbbox/Source Files/ViewModels/LikesViewModel.swift]: New line after a comma.

File: Inbbbox/Source Files/Managers/Animators/LoginViewAnimations.swift [hunk 7]
```diff
@@ -38,7 +38,7 @@ class LoginViewAnimations {
             UIView.animate(withDuration: 0.5, animations: { 
                 view.layer.cornerRadius = endValue
             })
-        }
+        }.catch { _ in }
```
Comment (Malibris) [inline Inbbbox/Source Files/Managers/Animators/LoginViewAnimations.swift]: Nothing. We do not base any logic on the result of this animation. So any error can be safely omitted (the result was unused before and generate a warning).

File: Inbbbox/Source Files/Providers/Core Data Providers/ManagedBlockedUsersRequester.swift [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        managedObjectsProvider = ManagedObjectsProvider(managedObjectContext: managedObjectContext)
+    }
+
+    func block(user: UserType) -> Promise<Void> {
+        return Promise<Void> { fulfill, reject in
+            if let _ = managedObjectsProvider.managedBlockedUser(user) {
+                reject(BlockedUserRequesterError.alreadyBlocked)
+            }
+
+            let managedUserEntity = NSEntityDescription.entity(forEntityName: ManagedBlockedUser.entityName, in: managedObjectContext)!
```
Comment (Malibris) [inline Inbbbox/Source Files/Providers/Core Data Providers/ManagedBlockedUsersRequester.swift]: Yup, `entity` returns an optional. I could use an optional binding though, but I followed the convention (see other `Requester` classes).

File: Inbbbox/Source Files/ViewControllers/BlockedUsersViewController.swift [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+// MARK: UITableViewDataSource
+
+extension BlockedUsersViewController {
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return users.count
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: BlockedUsersTableViewCell.identifier, for: indexPath) as! BlockedUsersTableViewCell
```
Comment (Malibris) [inline Inbbbox/Source Files/ViewControllers/BlockedUsersViewController.swift]: I did not know that, thanks. We should use it everywhere in the app.

File: Inbbbox/Source Files/ViewControllers/SettingsViewController.swift (hunk 1 above)
Comment (Malibris) [inline Inbbbox/Source Files/ViewControllers/SettingsViewController.swift]: Should be `if let destination = (item as? DetailsItem)?.destination {...`, but yes, I'll shorten it.

**Pull request merged**